libs/core/tests/unit_tests/test_tools.py

```diff
@@ -2174,3 +2175,146 @@ def foo(x: int) -> Bar:
+
+    class A:
+        def __init__(self, c: int):
+            self.c = c
+
+        @methodtool
+        def foo(self, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + self.c
+
+    a = A(10)
+    assert "self" not in a.foo.args
+
+
+async def test_method_tool_async() -> None:
+    """Test that a method tool can be async."""
+
+    class A:
+        def __init__(self, c: int):
+            self.c = c
+
+        @methodtool
+        async def foo(self, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + self.c
+
+    a = A(10)
+    async_response = await a.foo.ainvoke({"a": 1, "b": 2})
+    assert async_response == 13
+
+
+def test_method_tool_string_invoke() -> None:
+    """Test that a method tool can be invoked with a string."""
+
+    class A:
+        def __init__(self, a: str):
+            self.a = a
+
+        @methodtool
+        def foo(self, b: str) -> str:
+            """Concatenate a and b."""
+            return self.a + b
+
+    a = A("a")
+    assert a.foo.invoke("b") == "ab"
+
+
+def test_method_tool_toolcall_invoke() -> None:
+    """Test that a method tool can be invoked with a ToolCall."""
+
+    class A:
+        def __init__(self, c: int):
+            self.c = c
+
+        @methodtool
+        def foo(self, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + self.c
+
+    a = A(10)
+
+    tool_call = {
+        "name": a.foo.name,
+        "args": {"a": 1, "b": 2},
+        "id": "123",
+        "type": "tool_call",
+    }
+
+    tool_message = a.foo.invoke(tool_call)
+
+    assert int(tool_message.content) == 13
+
+
+def test_method_tool_info() -> None:
+    """Test that a method tools have correct name/description/args schema."""
+
+    class A:
+        def __init__(self, c: int):
+            self.c = c
+
+        @methodtool
+        def foo(self, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + self.c
+
+    args = {
+        "a": {"title": "A", "type": "integer"},
+        "b": {"title": "B", "type": "integer"},
+    }
+
+    assert A.foo.name == "foo"
+    assert A.foo.description == "Add two numbers to c."
+    assert A.foo.args == args
+
+
+def test_method_tool_classmethod() -> None:
+    """Test that a method tool can be a classmethod."""
+
+    class A:
+        c = 5
+
+        @methodtool
+        @classmethod
+        def foo(cls, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + cls.c
+
+    assert A.foo.invoke({"a": 1, "b": 2}) == 8
+
+
+def test_method_tool_classmethod_args() -> None:
+    """Test that a classmethod tool's args do not include cls."""
+
+    class A:
+        c = 5
+
+        @methodtool
+        @classmethod
+        def foo(cls, a: int, b: int) -> int:
+            """Add two numbers to c."""
+            return a + b + cls.c
+
+    assert "cls" not in A.foo.args
```

Comment:
```suggestion
    assert "cls" not in A.foo.args
    
class A:
    ...

class TestStandardAUnit(BaseToolUnitTests):
    ...
    
class TestStandardAIntegration(BaseToolIntegrationTests):
    ...
```

would also be good to implement some of these as standard tests - it tests a few of the exhaustive cases: https://python.langchain.com/docs/contributing/how_to/integrations/standard_tests/

sorry this is dragging on a bit! appreciate your help!

Reply:
This seems cool, I've never heard of this before and would like to give it a try since it could make the tests more concise. But I am having trouble putting them into the `test_tools.py` file. The issue comes from being unable to import the `ToolsUnitTests` class.